### PR TITLE
Fix scoring

### PR DESCRIPTION
**Pull request by Theodore-Chatziioannou**
Fixes some bugs of the PAM scoring module, validating them against MATSim output scores (from the "experienced plans" xml file).

Changes:
- the utility calculation for small durations is brought in line with the MATSim code [here](https://github.com/matsim-org/matsim-libs/blob/77536f9f05ff70b69bdf54f19604f5732d81949c/matsim/src/main/java/org/matsim/core/scoring/functions/CharyparNagelActivityScoring.java#L151-L193)
- fixed the "number of PT interchanges" calculation
- activity wrapping is only happening if the first and last activity are of the same type (see [here](https://github.com/matsim-org/matsim-libs/blob/77536f9f05ff70b69bdf54f19604f5732d81949c/matsim/src/main/java/org/matsim/core/scoring/functions/CharyparNagelActivityScoring.java#L241-L265)).
- boarding times can now be read from the experienced plans, enabling the calculation of waiting time disutilities
- waiting time (if any) is subtracted from the leg travel time
- conversion from `timedelta` to `hours` is done by dividing by a timedelta object (previously, `.seconds` were extracted, dropping any `day` elements, and resulting -for example- in erroneous activity duration for 24-hr activities).
- tests check the results against the MATSim-calculated scores of a Londinium sim
- PAM Plans now have a `.score` property, which gets populated with MATSim scores when reading experienced plans.

Next steps:
- align the scoring config format with the MC output json file, for easier interface between the two
- add some plan utility plots

### CODE REVIEW

**Comment by Theodore-Chatziioannou** [comment]
**fit against MATSim scores**

before:
scoring_before

after:
scoring_after_v4

File: pam/scoring.py [hunk 1]
```diff
@@ -146,18 +147,20 @@ def score_plan_activities(self, plan, cnfg):
         activities = list(plan.activities)
         if len(activities) == 1:
             return self.score_activity(activities[0], cnfg)
-        wrapped_activity, other_activities = self.activities_wrapper(activities)
-        return self.score_activity(wrapped_activity, cnfg) \
-            + sum([self.score_activity(act, cnfg) for act in other_activities if act.act != "pt interaction"])
-
+        if not activities[0].act == activities[-1].act:
```
Comment (ana-kop) [inline pam/scoring.py:150]: very minor but I would just write 

`if activities[0].act != activities[-1].act:`

File: pam/scoring.py [hunk 2]
```diff
@@ -177,22 +180,26 @@ def score_plan_legs(self, plan, cnfg):
...
-        for activity in plan.activities:
-            if activity.act == "pt_interaction":
-                in_transit += 1
-            else:
-                if in_transit:
+        for i in plan:
+            if isinstance(i, Activity):
+                if i.act not in ['pt interaction', 'pt_interaction']:
+                    if in_transit > 0:
+                        in_transit -= 1 # the first PT vehicle does not incur a line switch penalty
```
Comment (ana-kop) [inline pam/scoring.py:195]: Why are we subtracting 1 from `in_transit` rather than just keeping it as it was?

**Review by ana-kop — APPROVED** [review]
great improvements @Theodore-Chatziioannou , looks like significant move closer to MATSIM's scoring!

File: pam/scoring.py (hunk 1 above)
Comment (Theodore-Chatziioannou) [reply pam/scoring.py:150]: good point

File: pam/scoring.py (hunk 2 above)
Comment (Theodore-Chatziioannou) [reply pam/scoring.py:195]: so that the first PT vehicle boarding of each trip is not counted as a line switch.

**Review by fredshone — CHANGES_REQUESTED** [review]

File: pam/scoring.py (hunk 2 above)
Comment (fredshone) [reply pam/scoring.py:195]: 🤦🏻 I forgot about the walking. We havn't gone mad after all.

As things stand I belive we can take the number of pt interactions for a trip and subtract 2? As the start and end legs with always be access/egress interactions? However, that brings us close to the og implementation I think:

```
transits = []
in_transit = 0
for activity in plan.activities:
    if activity is a "pt_interaction":
        add one to transit counter
    else (must be a "proper" activity, ie not pt_interaction):
        append counter - 2 to transits list
        reset counter
total number of transits = sum of transits list
```

Comment (fredshone) [reply pam/scoring.py:195]: I am approving this as I do think your implementation is correct.

Although I am suspicious about:
```
for each trip:
    number of line switches = floor(number of pt interactions / 2) - 1
```

Comment (Theodore-Chatziioannou) [reply pam/scoring.py:195]: subtracting 2 not working, become _every_ alighting was a PT interaction, not just the last one.

Here is an example from Londinium:
```
<activity type="gym" link="5221366961124175583_5221366963840338509" x="527449.0377700026" y="174907.7934215893" start_time="10:47:41" end_time="12:10:44" >
<leg mode="walk" dep_time="12:10:44" trav_time="00:02:41">
<activity type="pt interaction" link="5221366961597271735_5221366961597271735" x="527518.5209884992" y="174830.5783811345" start_time="12:13:25" end_time="12:13:26" >
<leg mode="bus" dep_time="12:13:26" trav_time="00:08:34">
<activity type="pt interaction" link="4237" x="527375.7371101471" y="175397.20261675102" start_time="12:22:00" end_time="12:22:01" >
<leg mode="walk" dep_time="12:22:01" trav_time="00:05:09">
<activity type="pt interaction" link="5221366977608376877_5221366977608376877" x="527201.7595597449" y="175500.23172701587" start_time="12:27:10" end_time="12:27:11" >
<leg mode="rail" dep_time="12:27:11" trav_time="00:11:52">
<activity type="pt interaction" link="5221377924493935121_5221377924493935121" x="525381.6323965725" y="178000.1913795474" start_time="12:39:03" end_time="12:39:04" >
<leg mode="walk" dep_time="12:39:04" trav_time="00:00:29">
<activity type="pt interaction" link="5221377923036634347_5221377923036634347" x="525380.6561217004" y="178023.08433860267" start_time="12:39:33" end_time="12:39:34" >
<leg mode="subway" dep_time="12:39:34" trav_time="00:04:28">
<activity type="pt interaction" link="5221377909342570047_5221377909342570047" x="525521.6932877643" y="178560.11270567012" start_time="12:44:02" end_time="12:44:03" >
<leg mode="walk" dep_time="12:44:03" trav_time="00:09:49">
<activity type="home" link="4624" x="525833.7741580525" y="178773.8747479139" start_time="12:53:52" >
``` 

This trip has 6/2-1=2 interchanges, not 6-4=2.

Comment (fredshone) [reply pam/scoring.py:195]: checks out. I have been looking through some example outputs and these has been the case so far. I also understand why you are unsure - if there is a bus to bus interchange, using the same stop then perhaps there will not be a walk between.

I will see if i can check with columbus.

Comment (Theodore-Chatziioannou) [reply pam/scoring.py:195]: exactly

Comment (Theodore-Chatziioannou) [reply pam/scoring.py:195]: can you approve to merge then?

Comment (fredshone) [reply pam/scoring.py:195]: best to add a test...

```
	<person id="Jim">
		<attributes>
			<attribute name="hid" class="java.lang.String">Jim</attribute>
			<attribute name="subpopulation" class="java.lang.String">electric</attribute>
		</attributes>
		<plan score="85.96536618601868" selected="no">
			<activity type="home" link="1-1" x="0.0" y="5100.0" end_time="07:25:00" >
			</activity>
			<leg mode="walk" dep_time="07:25:00" trav_time="00:02:36">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="generic" start_link="1-1" end_link="1-1" trav_time="00:02:36" distance="130.0"></route>
			</leg>
			<activity type="pt interaction" link="1-1" x="0.0" y="5000.0" max_dur="00:00:00" >
			</activity>
			<leg mode="bus" dep_time="07:27:36" trav_time="00:12:24">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="default_pt" start_link="1-1" end_link="2-2" trav_time="00:12:24" distance="5001.0">{"transitRouteId":"city_bus","boardingTime":"07:30:00","transitLineId":"city_bus","accessFacilityId":"A","egressFacilityId":"B"}</route>
			</leg>
			<activity type="pt interaction" link="2-2" x="5000.0" y="5000.0" max_dur="00:00:00" >
			</activity>
			<leg mode="bus" dep_time="07:40:01" trav_time="00:10:59">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="default_pt" start_link="2-2" end_link="3-3" trav_time="00:10:59" distance="5001.0">{"transitRouteId":"city_bus2","boardingTime":"07:41:00","transitLineId":"city_bus2","accessFacilityId":"B","egressFacilityId":"F"}</route>
			</leg>
			<activity type="pt interaction" link="3-3" x="10100.0" y="5100.0" max_dur="00:00:00" >
			</activity>
			<leg mode="walk" dep_time="07:51:00" trav_time="00:00:00">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="generic" start_link="3-3" end_link="2-3" trav_time="00:00:00" distance="0.0"></route>
			</leg>
			<activity type="work" link="2-3" x="10100.0" y="5100.0" end_time="16:30:00" >
			</activity>
			<leg mode="walk" dep_time="16:30:00" trav_time="00:00:00">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="generic" start_link="2-3" end_link="3-3" trav_time="00:00:00" distance="0.0"></route>
			</leg>
			<activity type="pt interaction" link="3-3" x="10100.0" y="5100.0" max_dur="00:00:00" >
			</activity>
			<leg mode="bus" dep_time="16:30:00" trav_time="00:11:00">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="default_pt" start_link="3-3" end_link="2-2" trav_time="00:11:00" distance="5001.0">{"transitRouteId":"city_bus2","boardingTime":"16:31:00","transitLineId":"city_bus2","accessFacilityId":"F","egressFacilityId":"B"}</route>
			</leg>
			<activity type="pt interaction" link="2-2" x="5000.0" y="5000.0" max_dur="00:00:00" >
			</activity>
			<leg mode="bus" dep_time="16:41:01" trav_time="00:48:59">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="default_pt" start_link="2-2" end_link="1-1" trav_time="00:48:59" distance="5010.0">{"transitRouteId":"city_bus","boardingTime":"17:20:00","transitLineId":"city_bus","accessFacilityId":"B","egressFacilityId":"A"}</route>
			</leg>
			<activity type="pt interaction" link="1-1" x="0.0" y="5000.0" max_dur="00:00:00" >
			</activity>
			<leg mode="walk" dep_time="17:30:00" trav_time="00:02:36">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="generic" start_link="1-1" end_link="1-1" trav_time="00:02:36" distance="130.0"></route>
			</leg>
			<activity type="home" link="1-1" x="0.0" y="5100.0" >
			</activity>
		</plan>

		<plan score="77.762093585343" selected="yes">
			<activity type="home" link="1-1" x="0.0" y="5100.0" end_time="07:39:44" >
			</activity>
			<leg mode="walk" dep_time="07:39:44" trav_time="00:02:36">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="generic" start_link="1-1" end_link="1-1" trav_time="00:02:36" distance="130.0"></route>
			</leg>
			<activity type="pt interaction" link="1-1" x="0.0" y="5000.0" max_dur="00:00:00" >
			</activity>
			<leg mode="bus" dep_time="07:42:20" trav_time="00:27:40">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="default_pt" start_link="1-1" end_link="2-2" trav_time="00:27:40" distance="5001.0">{"transitRouteId":"city_bus","boardingTime":"08:00:00","transitLineId":"city_bus","accessFacilityId":"A","egressFacilityId":"B"}</route>
			</leg>
			<activity type="pt interaction" link="2-2" x="5000.0" y="5000.0" max_dur="00:00:00" >
			</activity>
			<leg mode="bus" dep_time="08:10:01" trav_time="00:40:59">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="default_pt" start_link="2-2" end_link="3-3" trav_time="00:40:59" distance="5001.0">{"transitRouteId":"city_bus2","boardingTime":"08:41:00","transitLineId":"city_bus2","accessFacilityId":"B","egressFacilityId":"F"}</route>
			</leg>
			<activity type="pt interaction" link="3-3" x="10100.0" y="5100.0" max_dur="00:00:00" >
			</activity>
			<leg mode="walk" dep_time="08:51:00" trav_time="00:00:00">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="generic" start_link="3-3" end_link="2-3" trav_time="00:00:00" distance="0.0"></route>
			</leg>
			<activity type="work" link="2-3" x="10100.0" y="5100.0" end_time="16:18:37" >
			</activity>
			<leg mode="walk" dep_time="16:18:37" trav_time="00:00:00">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="generic" start_link="2-3" end_link="3-3" trav_time="00:00:00" distance="0.0"></route>
			</leg>
			<activity type="pt interaction" link="3-3" x="10100.0" y="5100.0" max_dur="00:00:00" >
			</activity>
			<leg mode="bus" dep_time="16:18:37" trav_time="00:22:23">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="default_pt" start_link="3-3" end_link="2-2" trav_time="00:22:23" distance="5001.0">{"transitRouteId":"city_bus2","boardingTime":"16:31:00","transitLineId":"city_bus2","accessFacilityId":"F","egressFacilityId":"B"}</route>
			</leg>
			<activity type="pt interaction" link="2-2" x="5000.0" y="5000.0" max_dur="00:00:00" >
			</activity>
			<leg mode="bus" dep_time="16:41:01" trav_time="00:48:59">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="default_pt" start_link="2-2" end_link="1-1" trav_time="00:48:59" distance="5010.0">{"transitRouteId":"city_bus","boardingTime":"17:20:00","transitLineId":"city_bus","accessFacilityId":"B","egressFacilityId":"A"}</route>
			</leg>
			<activity type="pt interaction" link="1-1" x="0.0" y="5000.0" max_dur="00:00:00" >
			</activity>
			<leg mode="walk" dep_time="17:30:00" trav_time="00:02:36">
				<attributes>
					<attribute name="routingMode" class="java.lang.String">bus</attribute>
				</attributes>
				<route type="generic" start_link="1-1" end_link="1-1" trav_time="00:02:36" distance="130.0"></route>
			</leg>
			<activity type="home" link="1-1" x="0.0" y="5100.0" >
			</activity>
		</plan>

	</person>
```

Comment (fredshone) [reply pam/scoring.py:195]: ...let me know if you would like it reduced/simplified somehow

Comment (fredshone) [reply pam/scoring.py:195]: this is the output_experienced plan:
```
	<person id="Jim">
		<plan score="77.762093585343" selected="yes">
			<activity type="home" link="1-1" end_time="07:39:44" >
			</activity>
			<leg mode="walk" dep_time="07:39:44" trav_time="00:02:36">
				<route type="generic" start_link="1-1" end_link="1-1" trav_time="00:02:36" distance="130.0"></route>
			</leg>
			<activity type="pt interaction" link="1-1" x="0.0" y="5000.0" start_time="07:42:20" end_time="07:42:20" >
			</activity>
			<leg mode="bus" dep_time="07:42:20" trav_time="00:26:06">
				<attributes>
					<attribute name="enterVehicleTime" class="java.lang.Double">28801.0</attribute>
				</attributes>
				<route type="default_pt" start_link="1-1" end_link="2-2" trav_time="00:26:06" distance="5001.0">{"transitRouteId":"city_bus","boardingTime":"08:00:01","transitLineId":"city_bus","accessFacilityId":"A","egressFacilityId":"B"}</route>
			</leg>
			<activity type="pt interaction" link="2-2" x="5000.0" y="5000.0" start_time="08:08:26" end_time="08:08:26" >
			</activity>
			<leg mode="bus" dep_time="08:08:26" trav_time="00:40:57">
				<attributes>
					<attribute name="enterVehicleTime" class="java.lang.Double">31164.0</attribute>
				</attributes>
				<route type="default_pt" start_link="2-2" end_link="3-3" trav_time="00:40:57" distance="5001.0">{"transitRouteId":"city_bus2","boardingTime":"08:39:24","transitLineId":"city_bus2","accessFacilityId":"B","egressFacilityId":"F"}</route>
			</leg>
			<activity type="pt interaction" link="3-3" x="10100.0" y="5100.0" start_time="08:49:23" end_time="08:49:23" >
			</activity>
			<leg mode="walk" dep_time="08:49:23" trav_time="00:00:00">
				<route type="generic" start_link="3-3" end_link="2-3" trav_time="00:00:00" distance="0.0"></route>
			</leg>
			<activity type="work" link="2-3" x="10100.0" y="5100.0" start_time="08:49:23" end_time="16:18:37" >
			</activity>
			<leg mode="walk" dep_time="16:18:37" trav_time="00:00:00">
				<route type="generic" start_link="2-3" end_link="3-3" trav_time="00:00:00" distance="0.0"></route>
			</leg>
			<activity type="pt interaction" link="3-3" x="10100.0" y="5100.0" start_time="16:18:37" end_time="16:18:37" >
			</activity>
			<leg mode="bus" dep_time="16:18:37" trav_time="00:20:46">
				<attributes>
					<attribute name="enterVehicleTime" class="java.lang.Double">59401.0</attribute>
				</attributes>
				<route type="default_pt" start_link="3-3" end_link="2-2" trav_time="00:20:46" distance="5001.0">{"transitRouteId":"city_bus2","boardingTime":"16:30:01","transitLineId":"city_bus2","accessFacilityId":"F","egressFacilityId":"B"}</route>
			</leg>
			<activity type="pt interaction" link="2-2" x="5000.0" y="5000.0" start_time="16:39:23" end_time="16:39:23" >
			</activity>
			<leg mode="bus" dep_time="16:39:23" trav_time="00:44:10">
				<attributes>
					<attribute name="enterVehicleTime" class="java.lang.Double">60001.0</attribute>
				</attributes>
				<route type="default_pt" start_link="2-2" end_link="1-1" trav_time="00:44:10" distance="25014.0">{"transitRouteId":"city_bus","boardingTime":"16:40:01","transitLineId":"city_bus","accessFacilityId":"B","egressFacilityId":"A"}</route>
			</leg>
			<activity type="pt interaction" link="1-1" x="0.0" y="5000.0" start_time="17:23:33" end_time="17:23:33" >
			</activity>
			<leg mode="walk" dep_time="17:23:33" trav_time="00:02:36">
				<route type="generic" start_link="1-1" end_link="1-1" trav_time="00:02:36" distance="130.0"></route>
			</leg>
			<activity type="home" link="1-1" x="0.0" y="5100.0" start_time="17:26:09" >
			</activity>
		</plan>

	</person>
```

Comment (Theodore-Chatziioannou) [reply pam/scoring.py:195]: @fredshone I have now added the above test.

File: tests/test_12_scoring.py [hunk 3]
```diff
@@ -142,7 +142,11 @@ def test_scores_experienced(config_complex):
...
     for hid, pid, person in population.people():
-        person.attributes['subpopulation'] = 'default'
+        if 'subpopulation' not in person.attributes:
+            person.attributes['subpopulation'] = 'default'
         matsim_score = person.plan.score
         pam_score = scorer.score_person(person)
         assert abs(matsim_score - pam_score) < 0.1
+
+def test_score_pt_interchange(config_complex):
+    pass
```
Comment (fredshone) [inline tests/test_12_scoring.py:152]: pretty sketchy test

Comment (Theodore-Chatziioannou) [reply tests/test_12_scoring.py]: removed

**Review by fredshone — APPROVED** [review]

**Pull request merged**